### ISSUE TITLE
assert_eq! with a literal float should probably fire float_cmp_const not float_cmp

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust-clippy#6817]
Example:

```rust
pub const RADIANS_PER_FURMAN: f64 = f64::consts::PI / 32_768_f64;
pub const FURMANS_PER_RADIAN: f64 = 32_768_f64 / f64::consts::PI;
pub const DEGREES_PER_FURMAN: f64 = 45_f64 / 8192_f64;
pub const FURMANS_PER_DEGREE: f64 = 8192_f64 / 45_f64;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn constants_are_right() {
        // decimal expansions retrieved with WolframAlpha and truncated by clippy
        assert_eq!(RADIANS_PER_FURMAN, 0.000_095_873_799_242_852_57); // exact
        assert_eq!(FURMANS_PER_RADIAN, 10_430.378_350_470_453); // exact
        assert_eq!(DEGREES_PER_FURMAN, 0.005_493_164_062_5); // exact
        assert_eq!(FURMANS_PER_DEGREE, 182.044_444_444_444_45); // exact
    }
}
```

Here [`float_cmp`](https://rust-lang.github.io/rust-clippy/master/index.html#float_cmp) fires. Given that this is comparing a `const` to a literal float, however, this should almost certainly be [`float_cmp_const`](https://rust-lang.github.io/rust-clippy/master/index.html#float_cmp_const) instead (which is allow-by-default, as opposed to deny-by-default `float_cmp`).

This seems like a reasonable usage of exact floating point equality: defining `const`s with the intuitive definition and asserting that the produced value is precise within ±½ ULP, by checking against an alternatively produced value.

**Comment by constreference** [comment]
Hi, I need some help with this, I see that the is_named_constant in clippy_lints/src/misc.rs returns false, I am guessing this is due to the difference in expression objects between a regular variable and an assert_eq macro, could someone point me in the right direction to tackle this

**Comment by constreference** [comment]
CC @flip1995

**Comment by flip1995** [comment]
Hmm. I would start to add `dbg!()` around the `left` and `right` expression passed to `is_named_constant` in `misc` and then see what expressions get passed to `constant`. 

The `constant` function uses something from the `rustc` API directly. Maybe this has something to do with the macro context? But that would surprise me, because MIR shouldn't care about where constants come from.

**Comment by constreference** [comment]
This is how the debug expressions look like `   hir_id: HirId {
        owner: DefId(0:14 ~ float_cmp[317d]::main),
        local_id: 457,
    },
    kind: Unary(
        Deref,
        Expr {
            hir_id: HirId {
                owner: DefId(0:14 ~ float_cmp[317d]::main),
                local_id: 456,
            },
            kind: Path(
                Resolved(
                    None,
                    Path {
                        span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:23: 39:31 (#7),
                        res: Local(
                            HirId {
                                owner: DefId(0:14 ~ float_cmp[317d]::main),
                                local_id: 451,
                            },
                        ),
                        segments: [
                            PathSegment {
                                ident: left_val#7,
                                hir_id: Some(
                                    HirId {
                                        owner: DefId(0:14 ~ float_cmp[317d]::main),
                                        local_id: 455,
                                    },
                                ),
                                res: Some(
                                    Local(
                                        HirId {
                                            owner: DefId(0:14 ~ float_cmp[317d]::main),
                                            local_id: 451,
                                        },
                                    ),
                                ),
                                args: None,
                                infer_args: true,
                            },
                        ],
                    },
                ),
            ),
            attrs: ThinVec(
                None,
            ),
            span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:23: 39:31 (#7),
        },
    ),
    attrs: ThinVec(
        None,
    ),
    span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:22: 39:31 (#7),
}
[clippy_lints/src/misc.rs:728] right = Expr {
    hir_id: HirId {
        owner: DefId(0:14 ~ float_cmp[317d]::main),
        local_id: 460,
    },
    kind: Unary(
        Deref,
        Expr {
            hir_id: HirId {
                owner: DefId(0:14 ~ float_cmp[317d]::main),
                local_id: 459,
            },
            kind: Path(
                Resolved(
                    None,
                    Path {
                        span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:36: 39:45 (#7),
                        res: Local(
                            HirId {
                                owner: DefId(0:14 ~ float_cmp[317d]::main),
                                local_id: 452,
                            },
                        ),
                        segments: [
                            PathSegment {
                                ident: right_val#7,
                                hir_id: Some(
                                    HirId {
                                        owner: DefId(0:14 ~ float_cmp[317d]::main),
                                        local_id: 458,
                                    },
                                ),
                                res: Some(
                                    Local(
                                        HirId {
                                            owner: DefId(0:14 ~ float_cmp[317d]::main),
                                            local_id: 452,
                                        },
                                    ),
                                ),
                                args: None,
                                infer_args: true,
                            },
                        ],
                    },
                ),
            ),
            attrs: ThinVec(
                None,
            ),
            span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:36: 39:45 (#7),
        },
    ),
    attrs: ThinVec(
        None,
    ),
    span: /Users/srenganathan/.rustup/toolchains/nightly-2021-02-25-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:39:35: 39:45 (#7),
}`

**Comment by flip1995** [comment]
Ah this makes sense. Looking at the macro expansion:

```rust
        match (&RADIANS_PER_FURMAN, &0.000_095_873_799_242_852_57) {
            (left_val, right_val) => {
                if !(*left_val == *right_val) {
                    let kind = ::core::panicking::AssertKind::Eq;
                    ::core::panicking::assert_failed(kind, &*left_val,
                                                     &*right_val,
                                                     ::core::option::Option::None);
                }
            }
        }
```

Kind of weird, that rustc can't figure out that `right_val` is a constant. Something seems to go wrong in 

https://github.com/rust-lang/rust-clippy/blob/63aca96e9143bb03cc6f657aade80fdedc02e015/clippy_utils/src/consts.rs#L327-L328

Can you verify that this function gets called, while evaluating the constants? And possibly what the result of this is:

https://github.com/rust-lang/rust-clippy/blob/63aca96e9143bb03cc6f657aade80fdedc02e015/clippy_utils/src/consts.rs#L339-L350

**Comment by constreference** [comment]
This function indeed gets called, left and right expressions

```[clippy_utils/src/consts.rs:329] "Fetch path called" = "Fetch path called"
[clippy_utils/src/consts.rs:331] res = Local(
    HirId {
        owner: DefId(0:14 ~ float_cmp[317d]::main),
        local_id: 451,
    },
)
[clippy_utils/src/consts.rs:329] "Fetch path called" = "Fetch path called"
[clippy_utils/src/consts.rs:331] res = Local(
    HirId {
        owner: DefId(0:14 ~ float_cmp[317d]::main),
        local_id: 452,
    },
)```

For a normal expression (==) the value of result is as follows
```[clippy_utils/src/consts.rs:331] res = Def(
    Const,
    DefId(0:22 ~ float_cmp[317d]::main::ABRA),
)```

**Comment by flip1995** [comment]
So the `fetch_path` function does not deal with `Res::Local`. So to fix this issue, you have to figure out how to evaluate a constant from a `Local`/`HirId` (if at all possible) and add this here:
https://github.com/rust-lang/rust-clippy/blob/63aca96e9143bb03cc6f657aade80fdedc02e015/clippy_utils/src/consts.rs#L359-L360

**Comment by constreference** [comment]
Thanks, I'll look into that

**Comment by constreference** [comment]
@rustbot unassign
I'll not be able to look at this for the next 15 days, unassigning if anyone wants to take it up in the mean time

**Comment by torfsen** [comment]
> Here [float_cmp](https://rust-lang.github.io/rust-clippy/master/index.html#float_cmp) fires. Given that this is comparing a const to a literal float, however, this should almost certainly be [float_cmp_const](https://rust-lang.github.io/rust-clippy/master/index.html#float_cmp_const) instead (which is allow-by-default, as opposed to deny-by-default float_cmp).

As far as I can see, `float_cmp` is currently in the `pedantic` category, which is allow-by-default (`float_cmp_const` is in `restriction`, which is also allow-by-default). It was moved there in #7692 and released in [Rust 1.57](https://github.com/rust-lang/rust-clippy/blob/99c783843dbbdc0ff1af984666341544c962845a/CHANGELOG.md#rust-157). So changing the emitted lint would not create much of a difference here, at this point.

**Comment by CAD97** [comment]
Yeah, that's changed since I originally filed the issue. I updated the OP to not be incorrect.

Adjusting which lint gets emitted would still be more correct, but yes, it's significantly less impactful of a change with both lints being allow by default now. But it does still matter for when one but not the other is enabled.
